_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,220 @@
+  StartedOn   *time.Time   `gorm:"nullable;default:NULL" json:"startedOn" format:"date-time"`
+  CompletedOn *time.Time   `gorm:"nullable;default:NULL" json:"finishedOn" format:"date-time"`
+  GitBranch   string       `gorm:"size:300;default:'';not null" json:"gitBranch"`
+  Environment null.String  `gorm:"nullable;size:40" json:"environment" swaggertype:"string"`
+  Stage       string       `gorm:"size:40;default:'';not null" json:"stage"`
+  Params      []BuildParam `gorm:"foreignKey:BuildID" json:"params"`
+  IsInvalid   bool         `gorm:"not null;default:false" json:"isInvalid"` 
++ TestResultSummaries []TestResultSummary `gorm:"foreignKey:BuildID" json:"testResultSummaries"`
+}
+
++type TestResultSummary struct {
++  ArtifactID  uint      `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
++  Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Total       uint      `gorm:"not null" json:"total"`
++  Failed      uint      `gorm:"not null" json:"failed"`
++  Passed      uint      `gorm:"not null" json:"passed"`
++  Skipped     uint      `gorm:"not null" json:"skipped"`
++}
+
++type TestResult struct {
++  ArtifactID  uint              `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact         `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Name        string	        `gorm:"not null" json:"name"`
++  Message     string            `gorm:"nullable" json:"message"`
++  StartedOn   *time.Time        `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
++  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"finishedOn" format:"date-time"`
++  Status      TestResultStatus  `gorm:"not null" enums:"Failed,Passed,Skipped"`
++}
+```
+```go
+// artifact.go
+// new
+type File struct {
+  name string
+  fileName string
+  data []bytes
+}
+// new, /build/{buildid}/test-results
+func (m ArtifactModule) putTestResultsHandler(c *gin.Context) {
+  // parseMultipartFormData to get the files
+  // for each file, parse and store in db
+}
+// new, /build/{buildid}/artifact/{artifactid}/test-results
+func (m ArtifactModule) getBuildArtifactTestResultsHandler(c *gin.Context) {
+  // fetch test results for the specified test
+}
+// new
+func parseMultipartFormData(c *gin.Context) []*File {
+  // ...
+}
+// new
+func parseTRX(file *File) []TestResult, TestResultSummary {
+  // ...
+}
+```
+
+The new endpoint `GET /build/{buildid}/test-results-summary` fetches
+all test result summaries for a build and constructs a json response
+that contains them, and has a summary of them, like:
+
+<details><summary>GET /build/42/test-results-summary</summary>
```

Comment:
No damage in having a lengthy RFC. Suggest using subheaders instead of `<details>`. If it's worth hiding then it's not worth including at all, IMO. (edge cases may apply)